include/deal.II/fe/fe_tools_extrapolate.templates.h

```diff
@@ -1164,94 +1183,92 @@ namespace FETools
-      Utilities::MPI::CollectiveMutex::ScopedLock lock(mutex, communicator);
-
-      // We pick a new tag in each round. Wrap around after 10 rounds:
-      const int mpi_tag =
-        Utilities::MPI::internal::Tags::fe_tools_extrapolate + round % 10;
-
-      // send data
-      std::vector<std::vector<char>>::iterator buffer = sendbuffers.begin();
-      unsigned int                             idx    = 0;
-      for (typename std::vector<CellData>::const_iterator it =
-             cells_to_send.begin();
-           it != cells_to_send.end();
-           ++it, ++idx)
-        {
-          destinations.push_back(it->receiver);
-
-          *buffer        = it->pack_data();
-          const int ierr = MPI_Isend(buffer->data(),
-                                     buffer->size(),
-                                     MPI_BYTE,
-                                     it->receiver,
-                                     mpi_tag,
-                                     communicator,
-                                     &requests[idx]);
-          AssertThrowMPI(ierr);
-
-          ++buffer;
-        }
+      // First figure out where we need to send stuff. Some of the cells
+      // in the input argument to this function might be destined for
+      // the same process, so we have to only look at the unique set of
+      // destinations:
+      std::set<types::subdomain_id> destinations;
+      for (const auto &cell : cells_to_send)
+        destinations.insert(cell.receiver);
+
+      // Then set up the send/receive operation. This is best done through
+      // the 'consensus algorithm' setup that is used for point-to-point
+      // communication of information in cases where we do not know up
+      // front which processes (and from how many processes) we have to
+      // expect information from.
+      const auto get_destinations = [&destinations]() {
+        return std::vector<unsigned int>(destinations.begin(),
+                                         destinations.end());
+      };
 
-      Assert(destinations.size() == cells_to_send.size(), ExcInternalError());
+      const auto create_request =
+        [&cells_to_send](const types::subdomain_id other_rank,
+                         std::vector<char> &       send_buffer) {
+          // We have to create a single std::vector<char> that encodes the
+          // information from all cells that are to be sent to 'other_rank'.
+          // Each of the objects we want to send has a 'pack_data()'
+          // function, so we can just create a vector of std::vector<char>
+          // into which we put these strings, and then use the serialization
+          // framework to get it all into one string.
+          std::vector<std::vector<char>> array_of_strings;
+          for (const auto &cell : cells_to_send)
+            if (cell.receiver == other_rank)
+              array_of_strings.emplace_back(cell.pack_data());
+
+          send_buffer = Utilities::pack(array_of_strings);
```

Comment:
Let's disable compression.

Reply:
Also a good suggestion. The bulk of that object is a FP array that can't be compressed well anyway.